docs/architecture/adr-111-nil-mempool.md

```diff
@@ -0,0 +1,239 @@
+* when it is time to _reap_ (as it is currently called) the mempool the answer will always be empty
+* the application running on the proposer validator will add transactions it received
+  using the appchains's own mechanism via `PrepareProposal`.
+
+## Alternative Approaches
+
+These are the alternatives known to date:
+
+1. Keep the current model. Useful for basic apps, but clearly suboptimal for applications
+   with their own mechanism to disseminate transactions and particular performance requirements.
+2. Provide more efficient general-purpose mempool implementations.
+   This is an ongoing effort, but will take some time (and R&D effort) to come up with
+   advanced mechanisms -- likely highly configurable -- which then will have to be thoroughly tested.
+3. A similar approach to this one ([ADR110][adr-110]) whereby the application-specific
+   mechanism directly interacts with CometBFT via a newly defined gRPC interface.
+
+## Decision
+
+TBD
+
+## Detailed Design
+
+What this ADR proposes can already be achieved with an unmodified CometBFT since `v0.37.x`,
+albeit with a complex, poor UX.
+This core of this proposal is to make some internal changes so it is clear an simple for app developers,
+thus improving the UX.
+
+#### `nil` Mempool
+
+We propose a new mempool implementation, called `nil` Mempool, that effectively disables all mempool functionality
+within CometBFT.
+The `nil` Mempool implements the `Mempool` interface in a very simple manner:
+
+*	`CheckTx(tx types.Tx) (*abcicli.ReqRes, error)`: returns `nil, ErrNotAllowed`
+*	`RemoveTxByKey(txKey types.TxKey) error`: returns `ErrNotFound`
+* `ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs`: returns `nil`
+* `ReapMaxTxs(max int) types.Txs`: returns `nil`
+*	`Lock()`: does nothing
+* `Unlock()`: does nothing
+*	`Update(...) error`: returns `nil`
+* `FlushAppConn() error`: returns `nil`
+*	`Flush()`: does nothing
+* `TxsAvailable() <-chan struct{}`: returns a closed channel
+* `EnableTxsAvailable()`: does nothing
+* `SetTxRemovedCallback(cb func(types.TxKey))`: does nothing
+* `Size() int` returns 0
+* `SizeBytes() int64` returns 0
+
+Upon startup, the `nil` mempool reactor will advertise no channels to the peer to peer layer.
+
+### Configuration
+
+We propose the following changes to the `config.toml` file:
+
+```toml
+[mempool]
+# The type of mempool for this CometBFT node to use.
+#
+# Valid types of mempools supported by CometBFT:
+# - "flood" : clist mempool with flooding gossip protocol (default)
+# - "nil"   : nil-mempool (app needs an alternative tx dissemination mechanism)
+type = "nil"
+```
+
+The config validation logic will be modified to add a new rule that rejects a configuration file if:
+
+* the mempool is set to `nil`
+* `create_empty_blocks`, in `consensus` section, is set to `false`.
+
+### Impacted Workflows
+
+* *Submitting a transaction*. Users are not to submit transactions via CometBFT's RPC.
+  `BroadcastTx*` RPC methods will fail.
```

Comment:
How is the impact of this ? From a client's perspective (eg. wallets) how do they determine if they can submit a tx via `BroadcastTx` or not, is that based on the mempool configuration ?

Reply:
We should offer a meaningful return method that this does not work. However, this endpoint is still open to attacks , needlessly perhaps so it can be disabled?

Reply:
I believe It might be possible to add a condition that remove this route from the RPC, not as easy as what we did with the new gRPC but I think it is doable and I would prefer that we enforce things more than have expectations that people will not use it (or abuse) in the scenario that it should not be available

Reply:
instead of denying the rpc, id propose a short term solution of fire and forget. users can still submit via 26657 but checktx will fire and potentially not wait for the response, or discard the response. The reason is that that cosmjs puts a hard requirement on using tendermints rpc instead of the apps, this could cause lots of breakage for wallets and other clients

Reply:
> is that based on the mempool configuration ?

Yes. Probably the error returned should be explicit about that _errored_ (e.g. "nil-mempool configured, which does not accept transactions")

> it can be disabled?
That is another option. Will mention it in the text

> The reason is that that cosmjs puts a hard requirement on using tendermints rpc instead of the apps, this could cause lots of breakage for wallets and other clients

Then they should not configure the nil-mempool until they've remove that requirement. If we return a silent response, the wallets, etc. will stay broken, as the behavior they're expecting won't materialize.
Don't forget that the proposal on this ADR is additive and _not_ the default config (the default is still the current mempool).

Reply:
I'm also against the silent responses and think we should return https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501

Reply:
> Then they should not configure the nil-mempool until they've removed that requirement.

+1

> I'm also against the silent responses and think we should return https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501

Agreed. Technically the mechanism by which transactions _should_ be submitted when the nil mempool is enabled is via an API exposed by the application, executing `CheckTx` prior to insertion in the mempool.

The Cosmos SDK has an opportunity here to set the standard for such an API.